src/GraphQL/Execution/ExecutionStrategy.cs

```diff
@@ -47,22 +47,49 @@ public virtual async Task<ExecutionResult> ExecuteAsync(ExecutionContext context
         /// </summary>
         protected abstract Task ExecuteNodeTreeAsync(ExecutionContext context, ObjectExecutionNode rootNode);
 
+        /// <summary>
+        /// Returns the root graph type for the execution -- for a specified schema and operation type.
+        /// </summary>
+        protected virtual IObjectGraphType GetOperationRootType(Document document, ISchema schema, Operation operation)
+        {
+            IObjectGraphType type;
+
+            switch (operation.OperationType)
+            {
+                case OperationType.Query:
+                    type = schema.Query;
+                    break;
+
+                case OperationType.Mutation:
+                    type = schema.Mutation;
+                    if (type == null)
+                        throw new InvalidOperationError("Schema is not configured for mutations").AddLocation(operation, document);
+                    break;
+
+                case OperationType.Subscription:
+                    type = schema.Subscription;
+                    if (type == null)
+                        throw new InvalidOperationError("Schema is not configured for subscriptions").AddLocation(operation, document);
+                    break;
+
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(operation), "Can only execute queries, mutations and subscriptions.");
+            }
+
+            return type;
+        }
+
         /// <summary>
         /// Builds the root execution node.
         /// </summary>
         protected virtual RootExecutionNode BuildExecutionRootNode(ExecutionContext context, IObjectGraphType rootType)
         {
-            var root = new RootExecutionNode(rootType)
+            var root = new RootExecutionNode(rootType, context.Operation.SelectionSet)
             {
                 Result = context.RootValue
             };
 
-            var fields = System.Threading.Interlocked.Exchange(ref context.ReusableFields, null) ?? new Fields();
-
-            SetSubFieldNodes(context, root, fields.CollectFrom(context, rootType, context.Operation.SelectionSet));
-
-            fields.Clear();
-            System.Threading.Interlocked.CompareExchange(ref context.ReusableFields, fields, null);
+            SetSubFieldNodes(context, root);
```

Comment:
By adding the `SelectionSet` to the `RootExecutionNode`, duplicate code can be removed.

Reply:
I didn't like that either.